swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java

```diff
@@ -408,6 +437,40 @@ private void deleteContractMetric(
     contractMetricRepository.flush();
   }
 
+  private void deleteMisalignedSubscriptionMeasurements(
+      SubscriptionEntity existingSubscription, SubscriptionEntity matchingUpdatedSubscription) {
+    var measurements =
+        existingSubscription.getSubscriptionMeasurements().stream()
+            .filter(
+                measurement ->
+                    !matchingUpdatedSubscription
+                        .getSubscriptionMeasurements()
+                        .contains(measurement))
+            .toList();
+    measurements.forEach(
+        measurement -> deleteSubscriptionMeasurement(existingSubscription, measurement));
+  }
+
+  private void deleteSubscriptionMeasurement(
+      SubscriptionEntity subscriptionEntity,
+      SubscriptionMeasurementEntity subscriptionMeasurementEntity) {
+    log.info("Deleting subscription_measurement: {}", subscriptionMeasurementEntity);
+    subscriptionMeasurementRepository.delete(subscriptionMeasurementEntity);
+    subscriptionEntity.removeMeasurement(subscriptionMeasurementEntity);
+    subscriptionMeasurementRepository.flush();
+  }
+
+  private boolean subscriptionMeasurementsEqual(
+      SubscriptionEntity existing, SubscriptionEntity updated) {
+    if (existing.getSubscriptionMeasurements().size()
+        != updated.getSubscriptionMeasurements().size()) {
+      return false;
+    }
+    return updated.getSubscriptionMeasurements().stream()
+        .toList()
+        .containsAll(existing.getSubscriptionMeasurements().stream().toList());
+  }
```

Comment:
The IDE was warning that using containsAll for a List could result in poor performance since it involves a linear search for each element - so O(n * m) complexity.  A HashSet instead averages O(1) complexity for its `.contains` call.  The overall complexity for `.containsAll` will be O(n+m) instead of O(n * m).

```suggestion
  private boolean subscriptionMeasurementsEqual(
      SubscriptionEntity existing, SubscriptionEntity updated) {
    int existingSize = existing.getSubscriptionMeasurements().size();
    int updatedSize = updated.getSubscriptionMeasurements().size();

    if (existingSize != updatedSize) {
      return false;
    }

    var updatedMeasurements = new HashSet<>(updated.getSubscriptionMeasurements());
    var existingMeasurements = new HashSet<>(existing.getSubscriptionMeasurements());

    return updatedMeasurements.containsAll(existingMeasurements);
  }
```

Reply:
I was spinning my wheels trying to get the HashSet.containsAll() to work but it always comes back false :( I figured its something to do with the .hashCode() method and I made some updates to that and still can't get it working.